File: src/main/java/org/c3lang/intellij/completion/C3CompletionContributor.java
package org.c3lang.intellij.completion;

import com.intellij.codeInsight.completion.CompletionContributor;
import com.intellij.codeInsight.completion.CompletionInitializationContext;
import com.intellij.codeInsight.completion.CompletionType;

import static com.intellij.patterns.PlatformPatterns.psiElement;

public final class C3CompletionContributor extends CompletionContributor
{
	public C3CompletionContributor()
	{
		var pattern = psiElement();

		extend(CompletionType.BASIC, pattern, FunctionCompletionContributor.INSTANCE);
		extend(CompletionType.BASIC, pattern, TypeCompletionContributor.INSTANCE);
		extend(CompletionType.BASIC, pattern, ImportCompletionContributor.INSTANCE);
		extend(CompletionType.BASIC, pattern, ConstCompletionContributor.INSTANCE);
		extend(CompletionType.BASIC, pattern, FaultCompletionContributor.INSTANCE);
		extend(CompletionType.BASIC, pattern, TailExprCompletionContributor.INSTANCE);
		extend(CompletionType.BASIC, pattern, InitializerListCompletionContributor.INSTANCE);
		//extend(CompletionType.BASIC, pattern, DocCommentCompletionContributor) TODO
		extend(CompletionType.BASIC, pattern, TopLevelCompletionContributor.INSTANCE);
	}

	@Override
	public void beforeCompletion(CompletionInitializationContext context)
	{
		// path
		context.setDummyIdentifier(CompletionExtensionsKt.DUMMY_IDENTIFIER);
	}
}

File: src/main/java/org/c3lang/intellij/completion/C3CompletionContributor.kt
package org.c3lang.intellij.completion

import com.intellij.codeInsight.completion.CompletionContributor
import com.intellij.codeInsight.completion.CompletionInitializationContext
import com.intellij.codeInsight.completion.CompletionType
import com.intellij.patterns.PlatformPatterns.psiElement

class C3CompletionContributor : CompletionContributor()
{
    init
    {
        val pattern = psiElement()/*.inside(C3ModuleDefinition::class.java)*/

        extend(CompletionType.BASIC, pattern, FunctionCompletionContributor)
        extend(CompletionType.BASIC, pattern, TypeCompletionContributor)
        extend(CompletionType.BASIC, pattern, ImportCompletionContributor)
        extend(CompletionType.BASIC, pattern, ConstCompletionContributor)
        extend(CompletionType.BASIC, pattern, FaultCompletionContributor)
        extend(CompletionType.BASIC, pattern, TailExprCompletionContributor)
        extend(CompletionType.BASIC, pattern, InitializerListCompletionContributor)
        extend(CompletionType.BASIC, pattern, DocCommentCompletionContributor)
        extend(CompletionType.BASIC, pattern, TopLevelCompletionContributor)
    }

    override fun beforeCompletion(context: CompletionInitializationContext)
    {
        // path
        context.dummyIdentifier = DUMMY_IDENTIFIER
    }
}